src/components/FileList.tsx

```diff
@@ -237,8 +336,116 @@ export class FileList extends React.Component<IFileListProps, IFileListState> {
     await this.addFile(...this.markedFiles.map(file => file.to));
   };
 
-  updateSelectedFile = (file: Git.IStatusFile | null): void => {
-    this.setState({ selectedFile: file });
+  selectOnlyOneFile = (file: Git.IStatusFile): void => {
+    this.setState({
+      selectedFiles: [file],
+      lastClickedFile: file,
+      selectedFileStatus: file.status
+    });
+  };
+
+  toggleFile = (file: Git.IStatusFile): void => {
+    if (file.status !== this.state.selectedFileStatus) {
+      this.selectOnlyOneFile(file);
+      return;
+    }
+
+    const fileStatus = this.state.selectedFiles.find(fileStatus =>
+      filesAreEqual(fileStatus, file)
+    );
+    if (!fileStatus) {
+      this.setState({
+        selectedFiles: [...this.state.selectedFiles, file],
+        lastClickedFile: file
+      });
+    } else {
+      this.setState({
+        selectedFiles: this.state.selectedFiles.filter(
+          fileStatus => !filesAreEqual(fileStatus, file)
+        ),
+        lastClickedFile: file
+      });
+    }
+  };
+
+  removeDuplicateSelectedFiles = (): void => {
+    const selectedFiles = this.state.selectedFiles.filter(
+      (file, index, arr) => index === arr.findIndex(f => filesAreEqual(f, file))
+    );
+    this.setState({ selectedFiles });
+  };
+
+  selectFiles = (files: Git.IStatusFile[]): void => {
+    this.setState(
+      {
+        selectedFiles: [...this.state.selectedFiles, ...files]
+      },
+      () => this.removeDuplicateSelectedFiles()
+    );
+  };
```

Comment:
Why did you call `removeDuplicateSelectedFiles` after setting the state?

It would better to ensure we have a correct state directly:

```suggestion
  selectFiles = (files: Git.IStatusFile[]): void => {
    this.setState(prevState => {
      return {
        selectedFiles: [...prevState.selectedFiles, ...files.filter(file => !prevState.selectedFiles.some(f => filesAreEqual(f, file))]
      }}
    );
  };
```